chore(android): improve gradle build fault tolerance

jquick-axway opened the pull request:
**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-27718
- https://jira.appcelerator.org/browse/TIMOB-27720

**Summary:**
- [TIMOB-27718](https://jira.appcelerator.org/browse/TIMOB-27718) Log build warnings if "res" files have invalid names.
  * Restores old Titanium behavior where "res" file names are automatically renamed if invalid.
  * Will log build warnings if invalid along with a suggestion on what to rename it to.
- [TIMOB-27720](https://jira.appcelerator.org/browse/TIMOB-27720) Titanium SDK build should favor "python2" tool if available.
  * Avoids SDK build failures on systems that have python 3 installed.
  * Note: We plan on removing our python usage in the future.
- Now setting Java/Kotlin compiler max heap size via gradle.
  * Avoids `OutOfMemory` exception when building very large projects.
  * Uses existing `android.javac.maxmemory` setting.
  * Uses existing `android.dx.maxmemory` setting.
- Improved "AndroidManifest.xml" fault tolerance for app builds.
  * Now filters out `xmlns:android` namespaces attributes from child XML elements. Can only be defined in root `<manifest/>` element. CLI used to wrongly inject them.

---
**Test:**
1. Create a Classic Titanium app project.
2. Unzip "Resources.zip" attached to [TIMOB-27718](https://jira.appcelerator.org/browse/TIMOB-27718) to project's "Resources" folder.
3. Copy below "tiapp.xml" settings to project.
4. Copy below "app.js" to project.
5. Build and run on Android.
6. Verify that you see 2 Appc logos and a gray button image with dark border onscreen.
7. Verify that you see the following build warnings in the log.
```
[WARN]  Invalid "res" file: Resources/android/images/res-mdpi/1-appc-dev-logo.jpg
[WARN]  - Names must contain only lowercase a-z, 0-9, or underscore.
[WARN]  - Names cannot start with a number.
[WARN]  - Titanium will rename to: _1_appc_dev_logo.jpg
[WARN]  Invalid "res" file: Resources/android/images/res-mdpi/ButtonHoloDark.9.png
[WARN]  - Names must contain only lowercase a-z, 0-9, or underscore.
[WARN]  - Titanium will rename to: buttonholodark.9.png
[WARN]  Invalid "res" file: Resources/android/images/res-mdpi/Logos/Appc-Transparent.png
[WARN]  - Files cannot be put into subdirectories.
[WARN]  - Names must contain only lowercase a-z, 0-9, or underscore.
[WARN]  - Titanium will rename to: logos_appc_transparent.png
```

tiapp.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ti:app xmlns:ti="http://ti.appcelerator.org">
	<android xmlns:android="http://schemas.android.com/apk/res/android">
		<manifest>
			<uses-permission xmlns:android="" android:name="android.permission.FOREGROUND_SERVICE"/>
			<uses-permission xmlns:android="http://schemas.android.com/apk/res/android" android:name="android.permission.WAKE_LOCK"/>
			<application xmlns:android="">
				<meta-data xmlns:android="" android:name="my.metadata" android:value="Hello World"/>
			</application>
		</manifest>
	</android>
</ti:app>
```

app.js
```javascript
var imagePath;
var window = Ti.UI.createWindow();
var scrollView = Ti.UI.createScrollView({
	layout: "vertical",
	scrollType: "vertical",
	showHorizontalScrollIndicator: false,
	shorVerticalScrollIndicator: true,
	backgroundColor: "gray",
});
{
	imagePath = "/Resources/images/1-appc-dev-logo.jpg";
	scrollView.add(Ti.UI.createLabel({ text: imagePath }));
	scrollView.add(Ti.UI.createImageView({ image: imagePath }));
}
{
	imagePath = "/Resources/images/Logos/Appc-Transparent.png";
	scrollView.add(Ti.UI.createLabel({ text: imagePath }));
	scrollView.add(Ti.UI.createImageView({ image: imagePath }));
}
{
	imagePath = "/Resources/images/ButtonHoloDark.9.png";
	scrollView.add(Ti.UI.createLabel({ text: imagePath }));
	scrollView.add(Ti.UI.createView({ backgroundImage: imagePath, width: "50%", height: "25%" }));
}
window.add(scrollView);
window.open();
```

jquick-axway committed:
feat(android): log build warnings if res files have invalid names

- Restored old behavior (before Titanium 9.0.0) where we used to rename invalid "res" files into valid names.
  * No longer post-fixing MD5 hash to res files anymore (which was the old behavior).
- TIMOB-27718: Now log a build warning for every "res" file name that is invalid.

jquick-axway committed:
chore(android): ignore "xmlns:android" declarations in child elements

- Improve "AndroidManifest.xml" file reading fault tolerance for app builds.
  * CLI used to wrongly inject "xmlns:android" namespace declaration into child elements when updating "tiapp.xml". This works-around issue.

jquick-axway committed:
Merge branch 'master' into TIMOB-27718

jquick-axway committed:
chore(android): set java max heap size in gradle

- Re-added ability to set Java/Dexer max heap size for gradle builds.
  * Defaults to 3 GB like previous Titanium versions.
  * Now sets Kotlin build max heap size too.
  * Helps avoid OutOfMemory exceptions when doing app/module builds.

jquick-axway committed:
chore(android): favor "python2" tool for SDK builds

- TIMOB-27720: The "python" command line tool will use python 3 if installed, causing an SDK build failure.
  * SDK build system now checks if "python2" is installed use it if available.
  * This can be an issue on Linux and macOS.

jquick-axway committed:
Merge branch 'master' into TIMOB-27718

garymathews commented on android/cli/lib/android-manifest.js:
```diff
@@ -598,20 +598,30 @@ class AndroidManifest {
+			// Remove "xmlns:android" namespace attributes from all child elements. Only supported in <manifest/>.
+			// Note: CLI used to wrongly inject these after updating "tiapp.xml" if missing from root element.
+			const manifestEndIndex = text.indexOf('>', manifestIndex);
+			text = text.replace(/xmlns:android=".*?"/g, (match, p1) => {
```
In the regex `/xmlns:android=".*?"/g` is the `?` necessary, wouldn't the `.*` suffice?

jquick-axway replied:
Yes, the `?` is needed. Without it, the regex goes past the attribute and to the last double quote before the next match.

garymathews commented on android/cli/commands/_build.js:
```diff
@@ -898,12 +898,9 @@ AndroidBuilder.prototype.validate = function validate(logger, config, cli) {
-	this.javacSource = cli.tiapp.properties['android.javac.source'] && cli.tiapp.properties['android.javac.source'].value || config.get('android.javac.source', '1.7');
-	this.javacTarget = cli.tiapp.properties['android.javac.target'] && cli.tiapp.properties['android.javac.target'].value || config.get('android.javac.target', '1.7');
 	this.dxMaxMemory = cli.tiapp.properties['android.dx.maxmemory'] && cli.tiapp.properties['android.dx.maxmemory'].value || config.get('android.dx.maxMemory', '3072M');
-	this.dxMaxIdxNumber = cli.tiapp.properties['android.dx.maxIdxNumber'] && cli.tiapp.properties['android.dx.maxIdxNumber'].value || config.get('android.dx.maxIdxNumber', '65536');
```
Do we want to remove the ability to set `maxIdxNumber`?

jquick-axway replied:
The `d8` command line tool doesn't appear to support this. I ran `--help` on it and didn't find an equivalent.

garymathews reviewed (approved):
CR: **PASS**

jquick-axway committed:
chore(android): optimized gradle app project generation

- The generateAppProject() function's Promise.all() tasks were most calling blocking sync functions. Changed them to async.

jquick-axway committed:
chore(android): added missing AndroidX dependencies

- Added direct dependencies to libraries:
  * androidx.drawerlayout
  * androidx.swiperefreshlayout
  * androidx.viewpager
- Above were inherited via core dependencies, but updating core dependencies in app's "build.gradle" can cause the above to be lost.
  * Using "com.google.android.material" dependency 1.1.0 caused a NoClassDef found for "SwipeRefreshLayout". Directly linking to this dependency solves the problem.

ssjsamir committed:
Merge branch 'master' into TIMOB-27718

ssjsamir reviewed (approved):
FR Passed: Tested using the test steps mentioned in the description above. 

**Test Environment**
```
MacOS Catalina: 10.15.1 beta
Xcode: 11.3
Java Version: 1.8.0_131
Android NDK: 18.1.5063045
Node.js: 10.16.3
"NPM":"4.2.15-1","CLI":"7.1.2-7"
Pixel xl (7.1.1) Sim
```

garymathews committed:
Merge branch 'master' into TIMOB-27718
